pkg/controllers/user/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -0,0 +1,352 @@
+		for _, t := range mcapp.Spec.Targets {
+			apps, err := m.appLister.List(t.ProjectName, set.AsSelector())
+			if err != nil {
+				return nil, err
+			}
+			appsToDelete = append(appsToDelete, apps...)
+		}
+	}
+
+	var g errgroup.Group
+	for _, app := range appsToDelete {
+		g.Go(func() error {
+			var appWorkloadNamespace string
+			if app != nil {
+				appWorkloadNamespace = app.Spec.TargetNamespace
+			}
+			if err := m.apps.DeleteNamespaced(app.Namespace, app.Name, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err := m.namespaces.Delete(appWorkloadNamespace, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (m *MCAppController) getAllAppsToDelete(mcAppName string) ([]*pv3.App, *v3.MultiClusterApp, error) {
+	// to get all apps, get all clusters first, then get all apps in all projects of all clusters
+	appsToDelete := []*pv3.App{}
+	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcAppName})
+	clusters, err := m.clusterLister.List("", labels.NewSelector())
+	if err != nil {
+		return appsToDelete, nil, err
+	}
+	for _, c := range clusters {
+		projects, err := m.projectLister.List(c.Name, labels.NewSelector())
+		if err != nil {
+			return appsToDelete, nil, err
+		}
+		for _, p := range projects {
+			apps, err := m.appLister.List(p.Name, set.AsSelector())
+			if err != nil {
+				return appsToDelete, nil, err
+			}
+			appsToDelete = append(appsToDelete, apps...)
+		}
+	}
+	return appsToDelete, nil, err
+}
+
+func (m *MCAppController) update(mcappToUpdate *v3.MultiClusterApp) (*v3.MultiClusterApp, error) {
+	updatedObj, err := m.multiClusterApps.Update(mcappToUpdate)
+	if err != nil && apierrors.IsConflict(err) {
+		// retry 5 times
+		for i := 0; i < 5; i++ {
+			latestMcApp, err := m.multiClusterApps.GetNamespaced(globalNamespace, mcappToUpdate.Name, metav1.GetOptions{})
+			if err != nil {
+				return latestMcApp, err
+			}
+			latestToUpdate := latestMcApp.DeepCopy()
+			for ind, t := range mcappToUpdate.Spec.Targets {
+				if t.AppName != "" {
+					latestToUpdate.Spec.Targets[ind].AppName = t.AppName
+				}
+			}
+			updatedMcApp, err := m.multiClusterApps.Update(latestToUpdate)
+			if err != nil && apierrors.IsConflict(err) {
+				time.Sleep(5 * time.Millisecond)
+				continue
+			}
+			return updatedMcApp, err
+		}
+		return mcappToUpdate, err
+	}
+	return updatedObj, err
+}
+
+func (m *MCAppController) createAnswerMap(answers []v3.Answer) map[string]map[string]string {
+	var answerMap map[string]map[string]string
+
+	// create a map, where key is the projectID or clusterID, or "global" if neither is provided, and value is the actual answer values
+	answerMap = make(map[string]map[string]string)
+	for _, a := range answers {
+		if a.ProjectName != "" {
+			answerMap[a.ProjectName] = make(map[string]string)
```

Comment:
Should we check if answerMap[a.ProjectName] already exists and If yes just add to the existing map, else initialise a new map?

Reply:
I'm not sure that's needed, answerMap is a map and projectName would be a key, users shouldn't enter same key twice

Reply:
I'm not sure that's needed, while creating one instance of v3.Answer, users can provide projectName and shouldn't they provide all values for one project in one instance?